### PR TITLE
[buildingplan] allow filtering by yarn

### PR DESCRIPTION
**Pull request by myk002**
fixes https://github.com/DFHack/dfhack/issues/3251#issuecomment-1518139164

### CODE REVIEW

**Comment by TaxiService** [comment]
°-° does commenting count as reviewing? i hope so...

like, this is all fine and dandy as a temporary solution, but truthfully the behaviour of this filter selector does not look good. imo.

Selecting an item from the Type list should not automatically select *all* materials of that type. It should _filter_ the material list on the right, in the sense that: 
when a type is selected, only the materials of the selected type would be shown in the material list. 

Each of these lists should also include an "`any <materialtype>`" entry _at the top of the list_, and _that's_ what the user would click to select the whole category.
Doing so, would show just "any cloth" or "any leather" (or whatever the material is) in the `Current filter` panel/label.

despite requiring 1 additional click, this should make selecting a filter easier to read/comprehend. 
Right now the behaviour is unintuitive and the feedback too wordy.

this is an old mockup in which i had attempted to show what the type list should actually do.
In this image the types are different tabs, and clicking on each tab would switch to a new list for only that material.

All these new materials that popped in the Types list should've been put in the Other category.
When i made this though, i didn't think about letting the user choose individual types of leather/cloth/gems.
If we really want to provide such granularity, here are some ideas:
- we could simply have a tab for each of these categories, although it might get a little needlessly crowded.
- we could "merge" some categories together, such as `cloth`+`yarn`+`silk`->`textiles`, merge `glass` into `gem`, merge `clay` into `stone`, and anything else in `other`
- we could make some types "subcategories" of `other`, and have their tabs appear only after the `other` tab is selected...
- ...alternatively we could just have them all into `other` and have the first list entries be `any gem`, `any clay`, etc.

(this is an old mockup, but shows some concepts, such as having the 'any x', 'any y'... list entries at the top of the list, and the effect that this has visually on the `Current filter` panel.)
image

(also notice how selecting `any stone` puts a checkmark next to 'any stone' and a smaller green dot next to stones in the list, which are selected _indirectly_)
(the point is that the user _really_ doesn't need to see the whole list of minerals in the game if all they wanted was "any stone". :\ Hell, i'd rather have the filter text tell me if i have blocks/boulders/logs selected instead!)

**Review by TaxiService — APPROVED** [review]
°-° does commenting count as reviewing? i hope so...

like, this is all fine and dandy as a temporary solution, but truthfully the behaviour of this filter selector does not look good. imo.

Selecting an item from the Type list should not automatically select *all* materials of that type. It should _filter_ the material list on the right, in the sense that: 
when a type is selected, only the materials of the selected type would be shown in the material list. 

Each of these lists should also include an "`any <materialtype>`" entry _at the top of the list_, and _that's_ what the user would click to select the whole category.
Doing so, would show just "any cloth" or "any leather" (or whatever the material is) in the `Current filter` panel/label.

despite requiring 1 additional click, this should make selecting a filter easier to read/comprehend. 
Right now the behaviour is unintuitive and the feedback too wordy.

this is an old mockup in which i had attempted to show what the type list should actually do.
In this image the types are different tabs, and clicking on each tab would switch to a new list for only that material.

All these new materials that popped in the Types list should've been put in the Other category.
When i made this though, i didn't think about letting the user choose individual types of leather/cloth/gems.
If we really want to provide such granularity, here are some ideas:
- we could simply have a tab for each of these categories, although it might get a little needlessly crowded.
- we could "merge" some categories together, such as `cloth`+`yarn`+`silk`->`textiles`, merge `glass` into `gem`, merge `clay` into `stone`, and anything else in `other`
- we could make some types "subcategories" of `other`, and have their tabs appear only after the `other` tab is selected...
- ...alternatively we could just have them all into `other` and have the first list entries be `any gem`, `any clay`, etc.

(this is an old mockup, but shows some concepts, such as having the 'any x', 'any y'... list entries at the top of the list, and the effect that this has visually on the `Current filter` panel.)
image

(also notice how selecting `any stone` puts a checkmark next to 'any stone' and a smaller green dot next to stones in the list, which are selected _indirectly_)
(the point is that the user _really_ doesn't need to see the whole list of minerals in the game if all they wanted was "any stone". :\ Hell, i'd rather have the filter text tell me if i have blocks/boulders/logs selected instead!)

**Comment by myk002** [comment]
I fully agree that this is where we should go. I'm calling your proposed layout "buildingplan v2" and I intend to get it prioritized against the other items in the strategic roadmap. This PR is just solving a little nit in the current design so at least the functionality is there.

**Comment by myk002** [comment]
Oh, and to "approve" a review,

1. click "Files changed"
image

2. click "review changes"
image

3. leave comments, and select either the "comment" or "approve" radio button, then submit:
image

**Review by TaxiService — APPROVED** [review]
!!!!
